### PR TITLE
[[ Bug 15175 ]] Save the tabAlign property of fields in the stack file

### PR DESCRIPTION
**Pull request by livecodesebastien**

### CODE REVIEW

File: engine/src/field.cpp [hunk 1]
```diff
@@ -3205,11 +3207,28 @@ IO_stat MCField::extendedsave(MCObjectOutputStream& p_stream, uint4 p_part)
         t_size += sizeof(uint8_t);
     }
     
+    // SN-2015-04-30: [[ Bug 15175 ]] Save the tabalign property of the field
+    if (ntabs != 0)
+    {
+        t_flags |= FIELD_EXTRA_TABALIGN;
+        // Save number of tab alignments, and then each of them.
+        t_size += sizeof(uint16_t);
+        t_size += sizeof(intenum_t) * nalignments;
```
Comment (runrevmark) [inline engine/src/field.cpp]: There aren't enough alignments to justify using an intenum_t - the simplest thing to do is to serialize each alignment as bytes (although even that's a little overkill).

(The in-memory rep is abit overkill too!)

File: engine/src/field.cpp [hunk 2]
```diff
@@ -3262,8 +3312,9 @@ IO_stat MCField::save(IO_handle stream, uint4 p_part, bool p_force_ext)
 		return stat;
 
     // AL-2014-09-12: [[ Bug 13315 ]] Force an extension if field has explicit textDirection.
+    // SN-2015-04-30: [[ Bug 15175 ]] Force an extension if field has tabalign.
     bool t_has_extension;
-	t_has_extension = text_direction != kMCTextDirectionAuto;
+    t_has_extension = text_direction != kMCTextDirectionAuto || ntabs != 0;
```
Comment (runrevmark) [inline engine/src/field.cpp]: I think this should be nalignments != 0

**Comment by runrevmark** [comment]
Paragraphs can have the tabAlign property too :)

**Comment by livecodesebastien** [comment]
PR updated.

File: engine/src/field.cpp [hunk 3]
```diff
@@ -3205,11 +3207,28 @@ IO_stat MCField::extendedsave(MCObjectOutputStream& p_stream, uint4 p_part)
...
 
     if (t_stat == IO_NORMAL && (t_flags & FIELD_EXTRA_TEXTDIRECTION))
         t_stat = p_stream . WriteU8(text_direction);
+
+    if (t_stat == IO_NORMAL && (t_flags & FIELD_EXTRA_TABALIGN))
+    {
+        t_stat = p_stream . WriteU16(nalignments);
+
+        for (uint2 i = 0; i < nalignments && t_stat == IO_NORMAL; ++i)
+            t_stat = p_stream . WriteS8((int8_t)alignments[i]);
```
Comment (runrevmark) [inline engine/src/field.cpp]: I'd use p_stream . Write(alignments, nalignments) here - its a bit more efficient.

File: engine/src/field.cpp [hunk 4]
```diff
@@ -3238,6 +3257,37 @@ IO_stat MCField::extendedload(MCObjectInputStream& p_stream, uint32_t p_version,
...
+
+            if (t_stat == IO_NORMAL && t_nalign != 0)
+            {
+                if (!MCMemoryAllocate(sizeof(intenum_t) * t_nalign, t_alignments))
+                    t_stat = IO_ERROR;
+
+                for (uint2 i = 0; i < t_nalign && t_stat == IO_NORMAL; ++i)
+                {
+                    int8_t t_align;
+                    if ((t_stat = p_stream . ReadS8(t_align)) == IO_NORMAL)
```
Comment (runrevmark) [inline engine/src/field.cpp]: Similarly, here you can use Read(...) (to read the array all in one go).

File: engine/src/field.cpp (hunk 3 above)
Comment (runrevmark) [inline engine/src/field.cpp]: Ignore that - I misread the code.

File: engine/src/field.cpp (hunk 4 above)
Comment (runrevmark) [inline engine/src/field.cpp]: Ignore that - I misread the code.

**Comment by runrevmark** [comment]
Reviewed-by: @runrevmark.

**Comment by peter-b** [comment]
@livecode-vulcan review by @runrevmark ok bd50263

**Comment by livecode-vulcan** [comment]
💙 review by @runrevmark ok bd5026310b870a3d651b5e150ed085e2c6e35713

**Pull request merged**